### ISSUE TITLE
format! allows named arguments to be referenced by position

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/rust#69110]
The following example compiles, but shouldn't.
```rust
println!("{0} {foo}", foo = 2);
```

This seems to be an unintentional regression that was introduced somewhere between 1.11.0 and 1.12.0, most likely by #33642.

However, this code has been accepted on stable for multiple years now (1.12 was in 2016), so fixing this might break existing code. We could still change this back to an error if crater reports no regressions since it is reasonable to say that the code in question is clearly incorrect and was not caught by the compiler by mistake.

**Comment by estebank** [comment]
This should be reintroduced as a warn by default lint, regardless of crater results. We know this is relatively easy to stumble and rely on and rust is no longer close to 100% open source. Furthermore this is relatively benign mistake, a warning seems to me more than enough.

**Comment by ollie27** [comment]
This is pretty much a duplicate of #45256.

**Comment by pnkfelix** [comment]
> This is pretty much a duplicate of #45256.

(yeah but clearly no one acted on it back then, since even then the official response was "oh, we should lint against this.")

**Comment by Nemo157** [comment]
This has also leaked into the captured formatting parameter arguments in some situations (I could not come up with a situation where capturing a value parameter would do something similar):

```rust
fn main() {
    let a = 5;
    println!("{:-<a$}");
}
```

prints

```
5----
```

**Comment by Enselic** [comment]
Triage: #45256 has been closed as fixed, and the above code gives this warning now:

```
warning: named argument `foo` is not used by name
 --> src/main.rs:2:27
  |
2 |     println!("{0} {foo}", foo = 2);
  |               ---         ^^^ this named argument is referred to by position in formatting string
  |               |
  |               this formatting argument uses named argument `foo` by position
  |
  = note: `#[warn(named_arguments_used_positionally)]` on by default
help: use the named argument by name to avoid ambiguity
  |
2 |     println!("{foo} {foo}", foo = 2);
  |                ~~~

```

So we can close this issue too now.
